File: production-cms-backend-optimized/backend/src/utils/cache.js
```javascript
const redis = require('redis');
const config = require('../config');
const logger = require('./logger');

let client;

/**
 * Initializes the Redis client connection.
 * @returns {Promise<void>}
 */
const initRedis = async () => {
  if (client && client.isOpen) {
    logger.warn('Redis client already initialized and open.');
    return;
  }
  try {
    client = redis.createClient({
      url: config.redisUrl,
    });

    client.on('error', (err) => logger.error('Redis Client Error', err));
    client.on('connect', () => logger.info('Redis client connected.'));
    client.on('ready', () => logger.info('Redis client ready to use.'));
    client.on('end', () => logger.info('Redis client connection ended.'));

    await client.connect();
  } catch (err) {
    logger.error('Could not connect to Redis:', err);
    throw err; // Re-throw to indicate server startup failure
  }
};

/**
 * Gets a value from Redis cache.
 * @param {string} key - The cache key.
 * @returns {Promise<string|null>} The cached value or null.
 */
const get = async (key) => {
  if (!client || !client.isOpen) {
    logger.warn('Redis client is not connected. Skipping cache get.');
    return null;
  }
  try {
    return await client.get(key);
  } catch (err) {
    logger.error(`Error getting key ${key} from Redis:`, err);
    return null;
  }
};

/**
 * Sets a value in Redis cache with an optional expiration.
 * @param {string} key - The cache key.
 * @param {string} value - The value to cache.
 * @param {number} [expirationInSeconds] - Optional expiration time in seconds.
 * @returns {Promise<void>}
 */
const set = async (key, value, expirationInSeconds) => {
  if (!client || !client.isOpen) {
    logger.warn('Redis client is not connected. Skipping cache set.');
    return;
  }
  try {
    if (expirationInSeconds) {
      await client.setEx(key, expirationInSeconds, value);
    } else {
      await client.set(key, value);
    }
  } catch (err) {
    logger.error(`Error setting key ${key} in Redis:`, err);
  }
};

/**
 * Deletes a value from Redis cache.
 * @param {string} key - The cache key to delete.
 * @returns {Promise<void>}
 */
const del = async (key) => {
  if (!client || !client.isOpen) {
    logger.warn('Redis client is not connected. Skipping cache delete.');
    return;
  }
  try {
    await client.del(key);
  } catch (err) {
    logger.error(`Error deleting key ${key} from Redis:`, err);
  }
};

/**
 * Deletes all keys matching a pattern. Use with caution!
 * @param {string} pattern - The pattern to match (e.g., 'user:*').
 * @returns {Promise<void>}
 */
const delPattern = async (pattern) => {
  if (!client || !client.isOpen) {
    logger.warn('Redis client is not connected. Skipping cache delete by pattern.');
    return;
  }
  try {
    const keys = await client.keys(pattern);
    if (keys.length > 0) {
      await client.del(keys);
      logger.info(`Deleted ${keys.length} keys matching pattern: ${pattern}`);
    }
  } catch (err) {
    logger.error(`Error deleting keys by pattern ${pattern} from Redis:`, err);
  }
};


/**
 * Generates a consistent cache key from multiple arguments.
 * @param {...any} args - Arguments to form the key.
 * @returns {string} The generated cache key.
 */
const generateKey = (...args) => {
  return args.join(':');
};

module.exports = {
  initRedis,
  get,
  set,
  del,
  delPattern,
  generateKey,
  getClient: () => client, // For testing or direct access if needed
};
```

File: fullstack-realtime-chat-system-enhanced/backend/app/utils/cache.py
```python
import logging
import json
from typing import Any, Optional, Union

import redis.asyncio as aioredis
from redis.asyncio import Redis

from app.core.config import settings

logger = logging.getLogger(__name__)

class RedisClient:
    """
    A wrapper class for asynchronous Redis client operations.
    Provides methods for common caching patterns (get, set, delete).
    """
    def __init__(self):
        self._client: Optional[Redis] = None
        self.redis_url = settings.REDIS_URL
        logger.info(f"RedisClient initialized with URL: {self.redis_url}")

    async def init(self):
        """
        Initializes the asynchronous Redis client connection.
        This should be called during application startup.
        """
        if self._client is None:
            try:
                self._client = aioredis.from_url(self.redis_url, encoding="utf-8", decode_responses=True)
                await self._client.ping() # Test the connection
                logger.info("Redis client connected successfully.")
            except Exception as e:
                logger.error(f"Failed to connect to Redis at {self.redis_url}: {e}", exc_info=True)
                # In a real app, you might want to retry or raise a critical error
                raise ConnectionRefusedError(f"Could not connect to Redis: {e}") from e
        else:
            logger.info("Redis client already initialized.")

    @property
    def client(self) -> Redis:
        """
        Provides access to the underlying Redis client instance.
        Ensures init() has been called.
        """
        if self._client is None:
            raise RuntimeError("Redis client not initialized. Call .init() first.")
        return self._client

    async def close(self):
        """
        Closes the Redis client connection.
        This should be called during application shutdown.
        """
        if self._client:
            await self._client.close()
            logger.info("Redis client connection closed.")
            self._client = None

    async def get(self, key: str) -> Optional[Any]:
        """
        Retrieves a value from Redis. Assumes stored values are JSON strings.

        Args:
            key (str): The key to retrieve.

        Returns:
            Optional[Any]: The deserialized value, or None if key does not exist.
        """
        try:
            value = await self.client.get(key)
            if value:
                return json.loads(value)
            return None
        except Exception as e:
            logger.error(f"Error getting key '{key}' from Redis: {e}", exc_info=True)
            return None

    async def set(self, key: str, value: Any, ex: Optional[int] = None):
        """
        Stores a value in Redis. Serializes values to JSON strings.

        Args:
            key (str): The key to store the value under.
            value (Any): The value to store.
            ex (Optional[int]): Expiration time in seconds (e.g., 300 for 5 minutes).
        """
        try:
            serialized_value = json.dumps(value)
            await self.client.set(key, serialized_value, ex=ex)
            logger.debug(f"Set key '{key}' in Redis with expiry {ex}s.")
        except Exception as e:
            logger.error(f"Error setting key '{key}' in Redis: {e}", exc_info=True)

    async def delete(self, key: str) -> int:
        """
        Deletes one or more keys from Redis.

        Args:
            key (str): The key (or keys) to delete.

        Returns:
            int: The number of keys that were removed.
        """
        try:
            deleted_count = await self.client.delete(key)
            logger.debug(f"Deleted {deleted_count} keys from Redis: {key}")
            return deleted_count
        except Exception as e:
            logger.error(f"Error deleting key '{key}' from Redis: {e}", exc_info=True)
            return 0
    
    async def ping(self) -> bool:
        """
        Pings the Redis server to check connectivity.

        Returns:
            bool: True if connection is successful, False otherwise.
        """
        try:
            return await self.client.ping()
        except Exception as e:
            logger.error(f"Redis ping failed: {e}", exc_info=True)
            return False


redis_client = RedisClient()

```